### PR TITLE
fix(auth): add scheduler to schedule onAuth events through Angular zone

### LINKED ISSUE #354: fix(auth): observable not within angular zone
The auth observable, `af.auth`, is not in the angular zone when new auth objects are emitted. @StephenFluin has a reproduction in https://github.com/StephenFluin/ames

This is causing Angular's change detection not to run until the next tick.

### PR DESCRIPTION
**Pull request by jeffbcross**
Closes #354

### CODE REVIEW

File: src/auth/auth.spec.ts [hunk 1]
```diff
@@ -139,13 +139,17 @@ describe('FirebaseAuth', () => {
...
     authSpy['getRedirectResult'].and.returnValue(Promise.resolve(null));
 
-    inject([FirebaseApp, AngularFireAuth], (_app: firebase.app.App, _afAuth: AngularFireAuth) => {
-      app = _app;
-      afAuth = _afAuth;
-      authData = null;
-      authCb = null;
-      backend = new FirebaseSdkAuthBackend(app);
-    })();
+    var mockNgZone = Zone.current.fork({
```
Comment (vicb) [inline src/auth/auth.spec.ts]: Could you add a comment on why you need to wrap the code in a zone. What exactly needs to be wrapped ?

File: src/utils.ts [hunk 2]
```diff
@@ -75,4 +77,12 @@ export function stripLeadingSlash(value: string): string {
...
   }
-}
\ No newline at end of file
+}
+
+
+export class ZoneScheduler extends QueueScheduler {
+  constructor(_zone: Zone) {
+    super();
+    this.schedule = <(work: any) => Subscription>_zone.wrap(this.schedule, 'ZoneScheduler');
```
Comment (vicb) [inline src/utils.ts]: What about
- overriding the method ?
- creating a decorator that can wrap any scheduler ?

**Comment by vicb** [comment]
As discussed, please add a TODO comment to the code to fix the this properly. Is anyone working on a solution ?

File: src/auth/auth.spec.ts (hunk 1 above)
Comment (jeffbcross) [inline src/auth/auth.spec.ts]: This code actually doesn't need to be run in a zone. I should fork the zone in the spec that checks the zone to make sure the subscription callback isn't in the `<root>` zone.

**Comment by jeffbcross** [comment]
@vicb I'm working on addressing the bigger issue with Zone+Rx outside of this. But for now I changed the PR to wrap the subscriber functions at subscribe time.

File: src/auth/auth.spec.ts (hunk 1 above)
Comment (vicb) [inline src/auth/auth.spec.ts]: I mean an inline comment :)

File: src/auth/firebase_sdk_auth_backend.ts [hunk 3]
```diff
@@ -112,3 +111,10 @@ export class FirebaseSdkAuthBackend extends AuthBackend {
     }
   }
 }
+
+function wrapSubscriberMethods<T>(obs: Observer<T>): Observer<T> {
```
Comment (vicb) [inline src/auth/firebase_sdk_auth_backend.ts]: add a comment in the code

File: src/auth/firebase_sdk_auth_backend.ts [hunk 4]
```diff
@@ -112,3 +111,10 @@ export class FirebaseSdkAuthBackend extends AuthBackend {
...
+  obs.next = <(value: T) => void>Zone.current.wrap(obs.next, 'wrapSubscriberMethods');
```
Comment (vicb) [inline src/auth/firebase_sdk_auth_backend.ts]: `wrapSubscriberMethods.next`, ...

**Comment by vicb** [comment]
Please add comments to the code for maintainabilty.

LGTM once the comments are adressed

File: src/auth/auth.spec.ts (hunk 1 above)
Comment (jeffbcross) [inline src/auth/auth.spec.ts]: I know; I'm saying this code is not actually necessary.

**Comment by jeffbcross** [comment]
I'm going to make `zone` a required argument of `ZoneScheduler`, then will merge once green.

**Pull request merged**